walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,205 @@
+    List<TimeRange> optionalAttendeesEmptySlots = getEmptySlots(optionalAttendeesBusyTimeslots, meetingDuration);
+
+    if (mandatoryAttendees.size() == 0) {
+      return optionalAttendeesEmptySlots; 
+    } else if (optionalAttendees.size() == 0) {
+      return mandatoryAttendeesEmptySlots; 
+    } else {
+      List<TimeRange> allEmptySlots = findCommonEmptySlots(
+          mandatoryAttendeesEmptySlots, optionalAttendeesEmptySlots, meetingDuration);
+      return allEmptySlots;
+    }
+  }
+
+  /** 
+   * Constructs all the time ranges that any of the mandatory attendees is busy.
+   * These time ranges cannot be part of the potential meeting time. 
+   */ 
+  private List<TimeRange> getAllBusyTimeslots(Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+
+    events.forEach((event) -> {
+      Collection<String> eventAttendees = event.getAttendees();
+
+      for (String attendee : eventAttendees) {
+        // If the current event's attendee list contains a mandatory attendee, 
+        // adds this event's time range to the set of all busy times, 
+        // after checking for overlapping time ranges. 
+        if (mandatoryAttendees.contains(attendee)) {
+          TimeRange currEventTimeRange = event.getWhen(); 
+        
+          handleOverlapTimeRange(currEventTimeRange, busyTimes);
+
+          // If any one of the mandatory attendees is in this event's guest list,
+          // this event's time range needs to be considered busy. 
+          // Breaks out of the for-each loop because there is no need to look
+          // at the rest of the event's guest list. 
+          break; 
+        }
+      }
+    }); 
+    return busyTimes; 
+  }
+
+  /**
+   * Compares a given event's time slot with all the existing busy time slots.
+   * Merges the current event time slot with a busy time slot if they overlap.
+   * Adds the current event's time slot to the list of busy slots in the end. 
+   * @param current event's time range; a collection of busy time slots
+   */ 
+  private void handleOverlapTimeRange(TimeRange currEventTimeRange, Collection<TimeRange> busyTimes) {
+    boolean overlapped = false; 
+
+    for (TimeRange busyTimeRange : busyTimes) {
+      if (currEventTimeRange.overlaps(busyTimeRange)) {
+        TimeRange combinedTimeRange = getCombinedTimeRange(currEventTimeRange, busyTimeRange);
+        busyTimes.remove(busyTimeRange);
+        busyTimes.add(combinedTimeRange);
+        overlapped = true;  
+        currEventTimeRange = combinedTimeRange; 
+      } 
+    }
+
+    // If the current event doesn't overlap with any existing busy time range,
+    // add the current event's time range to the set of all busy time slots. 
+    if (! overlapped) {
+      busyTimes.add(currEventTimeRange); 
+    }
+  }
+
+  /** Creates a new overall combined time range for two time ranges that overlap */ 
```

Comment:
Can we have a period at the end of the sentence?

Reply:
Updated.